### PR TITLE
Add detailed_metric support for xds-config

### PR DESCRIPTION
**Pull request by akondapuram**
[Please NOTE] This PR has to be merged only after receiving approval for the [PR on the go-control-plane](https://github.com/envoyproxy/go-control-plane/pull/841). 
1. This PR adds support for detailed_metric field to dynamic ratelimiting in response to the [issue](https://github.com/envoyproxy/ratelimit/issues/461) I created a few weeks ago. 
2. The reason service RateLimitConfigDiscoveryService has been removed from this particular rls_conf.proto is because protoc would not compile it correctly. I think this was added after this file ([rls_conf.pb.go](https://github.com/envoyproxy/go-control-plane/blob/main/ratelimit/config/ratelimit/v3/rls_conf.pb.go)) was generated and checked-in to go-control-plane repo. 
3. Please see [commit](https://github.com/renuka-fernando/ratelimit/blob/df1e1b41bd571bb82a2ac794284c783fca9fa8e8/api/ratelimit/service/ratelimit/v3/rls_conf_ds.proto) for when rls_conf_ds.proto was initially added, but it was later removed as part of [another commit](https://github.com/envoyproxy/ratelimit/pull/373/commits/be6c1076fdce97718ac21c4eb3638f187674947a). 
4. In the [final commit ](https://github.com/envoyproxy/ratelimit/pull/373/commits/4f54625c2aadb3f9162b5dc129c6262931939698) rls_conf.proto was added along with service, but the corresponding [go-control-plane's files](https://github.com/envoyproxy/go-control-plane/tree/main/ratelimit) were not updated. Hence, removing it to reduce confusion here. (Another possible solution is to separate it out to another file (rls_conf_ds.proto) like before.

### CODE REVIEW

**Comment by dmi3zkm** [comment]
Hey @envoyproxy/ratelimit-maintainers
Could you please review this PR?

**Comment by birdayz** [comment]
bump :)
it would be really nice if we could merge this. the API part is merged into envoy's go-control-plane already.

**Comment by akondapuram** [comment]
@birdayz Though the PR is merged, currently awaiting the release of the new version of the go-control-plane repo. The latest version is from 4 weeks ago and doesn't have my changes yet. https://pkg.go.dev/github.com/envoyproxy/go-control-plane/ratelimit/config/ratelimit/v3

**Comment by birdayz** [comment]
Thanks for the quick response.
We could also depend on the commit `841e293a220b4d5abcc84ec0df41fd10a259ae42` directly (instead of using the tag), until the release of go-control-plane has happened. But i don't if that is against some policy of the ratelimit project.

**Comment by akondapuram** [comment]
@birdayz Thank you for the suggestion. I used commit instead of tag now and converted this PR from draft to review. 
@renuka-fernando Could you please review when you get some time? Thank you in advance.

File: api/ratelimit/config/ratelimit/v3/rls_conf.proto
Comment (renuka-fernando) [inline api/ratelimit/config/ratelimit/v3/rls_conf.proto:104]: We need the service definition here. Here [[1]](https://github.com/envoyproxy/go-control-plane/blob/29a4b731c77541ad7d5b31d3c7387820b2bda5b0/ratelimit/service/ratelimit/v3/rls_conf_ds.pb.go#L172) is the generated proto files and it is used in [[2]](https://github.com/envoyproxy/go-control-plane/blob/db2864461c76a66f05e55c31115dd43f296e9d95/pkg/server/v3/server.go#L56)

[1] https://github.com/envoyproxy/go-control-plane/blob/29a4b731c77541ad7d5b31d3c7387820b2bda5b0/ratelimit/service/ratelimit/v3/rls_conf_ds.pb.go#L172
[2] https://github.com/envoyproxy/go-control-plane/blob/db2864461c76a66f05e55c31115dd43f296e9d95/pkg/server/v3/server.go#L56

File: api/ratelimit/config/ratelimit/v3/rls_conf.proto
Comment (renuka-fernando) [inline api/ratelimit/config/ratelimit/v3/rls_conf.proto:94]: Let's keep the new line here.

**Comment by renuka-fernando** [comment]
Other changes LGTM.

File: api/ratelimit/config/ratelimit/v3/rls_conf.proto
Comment (akondapuram) [reply api/ratelimit/config/ratelimit/v3/rls_conf.proto:104]: Thank you for your review and comment @renuka-fernando 
As I mentioned in the point 4 of the [PR description](https://github.com/envoyproxy/ratelimit/pull/465#issue-2052970650), is it okay if I separate it out into another file rls_conf_ds.proto like how it was [here before](https://github.com/renuka-fernando/ratelimit/blob/df1e1b41bd571bb82a2ac794284c783fca9fa8e8/api/ratelimit/service/ratelimit/v3/rls_conf_ds.proto)? 
Also, the source of[ rls_conf_ds.pb.go](https://github.com/envoyproxy/go-control-plane/blob/29a4b731c77541ad7d5b31d3c7387820b2bda5b0/ratelimit/service/ratelimit/v3/rls_conf_ds.pb.go#L172) seems to be ratelimit/service/ratelimit/v3/rls_conf_ds.proto instead of the existing [rls_conf.proto](https://github.com/renuka-fernando/ratelimit/blob/main/api/ratelimit/config/ratelimit/v3/rls_conf.proto). 

Please see the following commits:
1. Both rls_conf.proto and rls_conf_ds.proto was added as part of [this commit ](https://github.com/envoyproxy/ratelimit/pull/373/commits/df1e1b41bd571bb82a2ac794284c783fca9fa8e8)
2. Later, they were removed as part of [this commit.](https://github.com/envoyproxy/ratelimit/pull/373/commits/be6c1076fdce97718ac21c4eb3638f187674947a#diff-04987d62d51285909050075cfd5a0567de1de1b7f7330c0c686347e4c7c477ee)
3. In the [final commit ](https://github.com/envoyproxy/ratelimit/pull/373/commits/4f54625c2aadb3f9162b5dc129c6262931939698)rls_conf.proto was added along with service, but the corresponding [go-control-plane's files](https://github.com/envoyproxy/go-control-plane/tree/main/ratelimit) were not updated.

Would like to hear your thoughts on how to proceed here. Thank you

Comment (renuka-fernando) [reply api/ratelimit/config/ratelimit/v3/rls_conf.proto:104]: Cool. Having a separate file would be fine. I can't exactly remember the reason for adding it to the same file.
Found this comment https://github.com/envoyproxy/go-control-plane/pull/598#discussion_r1008335478 referred in this commit: https://github.com/envoyproxy/ratelimit/pull/371/commits/0428a748fabb382e998af1e828d04bfb35c601f5

Yeah, separating it into another is fine.

**Review by renuka-fernando — APPROVED** [review]

**Comment by akondapuram** [comment]
@renuka-fernando Thank you for your review and approval.

**Comment by akondapuram** [comment]
@envoyproxy/ratelimit-maintainers Could you please help merge this PR? Thank you.

**Comment by ramaraochavali** [comment]
@mattklein123 @ysawa0 Can you please help merge this?

**Review by mattklein123 — APPROVED** [review]

**Comment by mattklein123** [comment]
Please check CI.

Also, is this a breaking change? It's a little unclear to me.

**Comment by akondapuram** [comment]
@mattklein123 Thank you for looking into this. It's not a breaking change, but this has dependency on the [go-control-plane ](https://github.com/envoyproxy/go-control-plane/pull/841) repo as we rely on rls_conf.pb.go file. This was merged a couple of weeks ago. Hence I kept this PR in draft until that was merged. Converted it to review after the other PR was merged.

Regarding the CI, Looks like there was a pre-commit end-of-file failure for rls_conf_ds.proto. Pushed a new commit. Could you please try to re-run the checks when you have some bandwidth? Thank you.

**Comment by akondapuram** [comment]
@mattklein123 Looks like all the checks are successful, could you please help merge this when you get some time? Thank you.

**Review by mattklein123 — APPROVED** [review]

**Comment by akondapuram** [comment]
@mattklein123 Thank you for your review and approval.

**Pull request merged**
